packages/hash/frontend/src/components/BlockLoader/data-map-editor.tsx

```diff
@@ -0,0 +1,357 @@
+import jsonpath from "jsonpath";
+import {
+  Dispatch,
+  ReactElement,
+  SetStateAction,
+  useMemo,
+  useState,
+} from "react";
+import { JsonSchema } from "@hashintel/hash-shared/json-utils";
+import { UnknownRecord } from "@blockprotocol/core";
+import { Box, Typography } from "@mui/material";
+import { faClose, faEye } from "@fortawesome/free-solid-svg-icons";
+import { BlockGraph, Entity } from "@blockprotocol/graph";
+import {
+  FontAwesomeIcon,
+  IconButton,
+  MenuItem,
+  Select,
+  TextField,
+} from "@hashintel/hash-design-system";
+import { dataTreeFromEntityGraph, SchemaMap } from "./shared";
+import { DataMapPreview } from "./data-map-editor/data-map-preview";
+
+type DataMapperProps = {
+  close: () => void;
+  mapId: string;
+  schemaMap?: SchemaMap;
+  sourceBlockEntity: Entity;
+  sourceBlockGraph: BlockGraph;
+  targetSchema: JsonSchema;
+  transformedTree: UnknownRecord;
+  updateSchemaMap: Dispatch<SetStateAction<SchemaMap>>;
+};
+
+/**
+ * Generates an array of strings, each of which represents a path in an object
+ * Where paths are nested, they are dot-separated, e.g. "address.city"
+ * These are jsonpaths with the leading "$." stripped for use in lodash methods
+ */
+const derivePathsFromObject = (object: {}) =>
+  jsonpath.paths(object, "$..*").map((path) => path.slice(1).join("."));
+
+/**
+ * Derive paths to a theoretical JSON object from a JSON schema and a set of properties
+ * Based on https://stackoverflow.com/a/60297881
+ * This does not handle $defs since we will not have them in the new type system.
+ */
+const derivePathsFromSchema = (
+  schema: JsonSchema,
+  schemaProperties: JsonSchema["properties"],
+  allPathsSoFar: string[] = [],
+): string[] => {
+  if (!schemaProperties) {
+    return allPathsSoFar;
+  }
+  return Object.entries(schemaProperties).reduce(
+    (pathAccumulator, [currentKey, currentSchema]) => {
+      if (!currentSchema) {
+        return pathAccumulator;
+      }
+      if (currentSchema.type === "object") {
+        return pathAccumulator.concat(
+          currentKey,
+          derivePathsFromSchema(
+            schema,
+            currentSchema.properties,
+            allPathsSoFar,
+          ).map((path) => `${currentKey}.${path}`),
+        );
+      }
+      if (
+        currentSchema.type === "array" &&
+        currentSchema.items &&
+        "properties" in currentSchema.items
+      ) {
+        return pathAccumulator.concat(
+          currentKey,
+          `${currentKey}[]`,
+          derivePathsFromSchema(
+            schema,
+            currentSchema.items.properties,
+            allPathsSoFar,
+          ).map((path) => `${currentKey}[].${path}`),
+        );
+      }
+
```

Comment:
I left this as-is because you if the `type` is `"array"`, you have to return one thing if it's an array of objects, and hit the following condition (just return the key) if it isn't. That means including an if statement inside the array case, one of which repeats the return in the default case, making it about the same clarity wise.

Reply:
You don't need to use `default` when your switch cases use `return`. You can have the default outside of the switch, and then have it fall through, removing need to duplicate. 

```typescript
switch (currentSchema.type) {
  case "object":
    return pathAccumulator.concat(
      currentKey,
      derivePathsFromSchema(
        schema,
        currentSchema.properties,
        allPathsSoFar,
      ).map((path) => `${currentKey}.${path}`),
    );
  case "array":
    if (currentSchema.items && "properties" in currentSchema.items) {
      return pathAccumulator.concat(
        currentKey,
        `${currentKey}[]`,
        derivePathsFromSchema(
          schema,
          currentSchema.items.properties,
          allPathsSoFar,
        ).map((path) => `${currentKey}[].${path}`),
      );
    }
    break;
}

return pathAccumulator.concat(currentKey);
```

It's up to you, I just find `switch` statements more readable.

I *think* you've maybe made a later change that invalidates this, checking on that now.

Reply:
I think they're more readable when the conditions you're checking line up neatly with the cases, but for me the fact that we have a handful of possible cases (object, array, string, number, null), of which 1.5 are handled in the switch statement and the rest outside negate any readability benefits that the syntax has.

Reply:
Ah yes, this is invalidated by https://github.com/hashintel/hash/commit/64599d48445127b121a8bfbe4010f7f7d76033d3 – ignore me.